include/stdx/span.hpp

```diff
@@ -0,0 +1,256 @@
+        return std::reverse_iterator{end()};
+    }
+    [[nodiscard]] constexpr auto crbegin() const noexcept
+        -> const_reverse_iterator {
+        return std::reverse_iterator{cend()};
+    }
+
+    [[nodiscard]] constexpr auto end() const noexcept -> iterator {
+        return ptr + this->size();
+    }
+    [[nodiscard]] constexpr auto cend() const noexcept -> const_iterator {
+        return ptr + this->size();
+    }
+    [[nodiscard]] constexpr auto rend() const noexcept -> reverse_iterator {
+        return std::reverse_iterator{begin()};
+    }
+    [[nodiscard]] constexpr auto crend() const noexcept
+        -> const_reverse_iterator {
+        return std::reverse_iterator{cbegin()};
+    }
+
+    [[nodiscard]] constexpr auto front() const -> reference { return *begin(); }
+    [[nodiscard]] constexpr auto back() const -> reference {
+        return *std::prev(end());
+    }
+
+    [[nodiscard]] constexpr auto operator[](size_type idx) const -> reference {
+        return data()[idx];
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto first() const -> span<element_type, Count> {
+        return span<element_type, Count>{ptr, Count};
+    }
+    [[nodiscard]] constexpr auto first(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr, count};
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto last() const -> span<element_type, Count> {
+        return span<element_type, Count>{ptr + this->size() - Count, Count};
+    }
+    [[nodiscard]] constexpr auto last(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr + this->size() - count, count};
+    }
+
+    template <std::size_t Offset, std::size_t Count = dynamic_extent>
+    [[nodiscard]] constexpr auto subspan() const {
+        if constexpr (Count != dynamic_extent) {
+            return span<element_type, Count>{ptr + Offset, Count};
+        } else if constexpr (Extent != dynamic_extent) {
+            return span<element_type, Extent - Offset>{ptr + Offset,
+                                                       Extent - Offset};
+        } else {
+            return span<element_type, dynamic_extent>{ptr + Offset,
+                                                      this->size() - Offset};
+        }
+    }
+
+    [[nodiscard]] constexpr auto subspan(size_type Offset,
+                                         size_type Count = dynamic_extent) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr + Offset, std::min(Count, this->size() - Offset)};
+    }
+};
+
+// NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast)
+template <class T, std::size_t N> auto as_bytes(span<T, N> s) noexcept {
+    if constexpr (N == dynamic_extent) {
+        return span{reinterpret_cast<std::byte const *>(s.data()),
+                    s.size_bytes()};
+    } else {
+        constexpr auto Size = sizeof(T) * N;
```

Comment:
**question**: `size_bytes` should be `constexpr` when `N != dynamic_extent`, no?

Reply:
Yes - that's simpler. Changed.